### ISSUE TITLE
sstable - writer fails/crashes if schema is altered during the write operation

### ISSUE THREAD

**Issue by enaydanov** [scylladb/scylladb#16065]
## Issue description

- [ ] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

There is a sequence of nemeses during longevity-150gb-asymmetric-cluster-12h-test which not failed actually, but produced errors during their run.

The main problematic nemesis is MgmtRestore during which we got following errors:

```
2023-11-11 11:21:45.080: (CassandraStressLogEvent Severity.ERROR) period_type=one-time event_id=a46d1e86-265b-4b66-aec3-5e64ed63da42 during_nemesis=MgmtRestore: type=OperationOnKey regex=Operation x10 on key\(s\) \[ line_number=662 node=Node longevity-50gb-12h-master-loader-node-df57e125-1 [54.154.200.67 | 10.4.8.30] (seed: False)
java.io.IOException: Operation x10 on key(s) [31383334393132383431]: Error executing: (UnavailableException): Not enough replicas available for query at consistency QUORUM (2 required but only 0 alive)
...
2023-11-11 11:22:36.841: (CassandraStressEvent Severity.ERROR) period_type=end event_id=a46d1e86-265b-4b66-aec3-5e64ed63da42 during_nemesis=MgmtRestore duration=1m2s: node=Node longevity-50gb-12h-master-loader-node-df57e125-1 [54.154.200.67 | 10.4.8.30] (seed: False)
stress_cmd=cassandra-stress read cl=QUORUM n=5242880 -schema 'keyspace=5gb_sizetiered_5_2 replication(strategy=NetworkTopologyStrategy,replication_factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -mode cql3 native  -rate threads=50 -col 'size=FIXED(64) n=FIXED(16)' -pop seq=1..5242880
errors:
Stress command completed with bad status 1: WARN  11:21:41,820 Error while computing token map for keyspace 5gb_sizetiered_5_2 with datacenter u
``` 

This problem is tracked here: https://github.com/scylladb/scylla-manager/issues/3525

But in the last run (https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-150gb-asymmetric-cluster-12h-test/28/) we got additional error during consequent ModifyTable nemesis:

```
2023-11-11 23:52:25.693 <2023-11-11 23:52:25.627>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=1adb6f8a-6998-4c5f-8274-308c87b2b574: type=RUNTIME_ERROR regex=std::runtime_error line_number=60146 node=longevity-50gb-12h-master-db-node-df57e125-3
2023-11-11T23:52:25.627+00:00 longevity-50gb-12h-master-db-node-df57e125-3      !ERR | scylla[35229]:  [shard  9:comp] sstable - writer failed to close file: std::runtime_error (compress is not supported)
```

This looks like a regression from `5.5.0~dev-20231027.227136ddf54f` but there is a difference in nemeses targets: in `5.5.0~dev-20231108.a4aeef2eb0aa` run the target is the same (node-1) but in `5.5.0~dev-20231027.227136ddf54f` MgmtRestore targeted at node-4 and ModifyTable 
at node-6

## Installation details

Kernel Version: 5.15.0-1049-aws
Scylla version (or git commit hash): `5.5.0~dev-20231108.a4aeef2eb0aa` with build-id `6c90f896983cf138cd171d49f9d406876333f910`

Cluster size: 4 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-50gb-12h-master-db-node-df57e125-6 (34.242.221.56 | 10.4.8.195) (shards: 14)
- longevity-50gb-12h-master-db-node-df57e125-5 (54.74.150.110 | 10.4.11.122) (shards: 8)
- longevity-50gb-12h-master-db-node-df57e125-4 (34.242.136.68 | 10.4.8.97) (shards: 10)
- longevity-50gb-12h-master-db-node-df57e125-3 (3.252.147.7 | 10.4.8.72) (shards: 14)
- longevity-50gb-12h-master-db-node-df57e125-2 (52.209.113.142 | 10.4.11.192) (shards: 14)
- longevity-50gb-12h-master-db-node-df57e125-1 (63.33.65.56 | 10.4.9.54) (shards: 8)

OS / Image: `ami-0c5cf87017b2b6cb8` (aws: undefined_region)

Test: `longevity-150gb-asymmetric-cluster-12h-test`
Test id: `df57e125-db71-4926-87d5-1867de178e42`
Test name: `scylla-master/longevity/longevity-150gb-asymmetric-cluster-12h-test`
Test config file(s):

- [longevity-150GB-12h-autorization-LimitedMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/6f6ce1209c6304330f85ffd16f12942c84db7d0d/test-cases/longevity/longevity-150GB-12h-autorization-LimitedMonkey.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor df57e125-db71-4926-87d5-1867de178e42`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=df57e125-db71-4926-87d5-1867de178e42)
- Show all stored logs command: `$ hydra investigate show-logs df57e125-db71-4926-87d5-1867de178e42`

## Logs:
- **db-cluster-df57e125.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/db-cluster-df57e125.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/db-cluster-df57e125.tar.gz)
- **sct-runner-events-df57e125.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/sct-runner-events-df57e125.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/sct-runner-events-df57e125.tar.gz)
- **sct-df57e125.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/sct-df57e125.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/sct-df57e125.log.tar.gz)
- **loader-set-df57e125.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/loader-set-df57e125.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/loader-set-df57e125.tar.gz)
- **monitor-set-df57e125.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/monitor-set-df57e125.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/df57e125-db71-4926-87d5-1867de178e42/20231112_051546/monitor-set-df57e125.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-150gb-asymmetric-cluster-12h-test/28/)
[Argus](https://argus.scylladb.com/test/6a3a708c-caa0-439f-84cd-736b40b517e3/runs?additionalRuns[]=df57e125-db71-4926-87d5-1867de178e42)
</details>

**Comment by mykaul** [comment]
@enaydanov  : 
1. Why is it a regression? I saw this in tests from 5.2 era. (email title: "[qa] FAILED: scylla-master/SCT_Raft_Experimental/longevity-lwt-parallel-schema-changes-with-disruptive-24h-with-raft-test#23: 2022-12-30 04:35:43" )
2. What's the user impact here? More specifically, did compaction succeed or not?

**Comment by enaydanov** [comment]
1) I described in the bug that during the previous run (# 27) we had same sequence of problems except this one. Also I'm not sure if longevity-lwt-parallel-schema-changes-with-disruptive-24h-with-raft-test#23 is the same problem because sequence of nemeses is completely different.

2) I don't see any impact except errors in the log.

In scylla.log it looks like following, but in 5 seconds (see "Will retry..." message) there is no more such error.

```
2023-11-11T23:52:25.627+00:00 longevity-50gb-12h-master-db-node-df57e125-3      !ERR | scylla[35229]:  [shard  8:comp] sstable - writer failed to close file: std::runtime_error (compress is not supported)
2023-11-11T23:52:25.627+00:00 longevity-50gb-12h-master-db-node-df57e125-3     !INFO | scylla[35229]:  [shard  8:comp] compaction - [Compact keyspace1.standard1 5af7fc70-80ed-11ee-812a-465d5608a31b] Compacting of 7 sstables interrupted due to: std::runtime_error (compress is not supported)
2023-11-11T23:52:25.627+00:00 longevity-50gb-12h-master-db-node-df57e125-3      !ERR | scylla[35229]:  [shard  8:comp] compaction_manager - Compaction task 0x60b006fe0000 for table keyspace1.standard1 compaction_group=0/1 [0x60b0037764a0]: failed: std::runtime_error (compress is not supported). Will retry in 5 seconds
```

in 5 secs:

```
2023-11-11T23:52:30.763+00:00 longevity-50gb-12h-master-db-node-df57e125-3     !INFO | scylla[35229]:  [shard  8:comp] compaction - [Compact keyspace1.standard1 602a1840-80ed-11ee-812a-465d5608a31b] Compacting [/ var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2n_5oh401ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c 95e2/me-3gaz_1u2i_18snk1ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2r_4bluo1ypnor6yvttsr-big-Data.db:level=3:origin=comp action,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2m_0qsi81ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08 d028cf1c95e2/me-3gaz_1u2k_17icw1ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2g_13v6o1ypnor6yvttsr-big-Data.db:level=3:ori gin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1svi_45e341ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c 11eeac08d028cf1c95e2/me-3gaz_1svg_2t5z41ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2t_3r8ts1ypnor6yvttsr-big-Data.db:lev el=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2e_067rk1ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5b fd30804c11eeac08d028cf1c95e2/me-3gaz_1tlb_1sq8w1ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2c_0saio1ypnor6yvttsr-big-Dat a.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1u2p_4pyts1ypnor6yvttsr-big-Data.db:level=3:origin=compaction,/var/lib/scylla/data/keyspace1/standa rd1-dc5bfd30804c11eeac08d028cf1c95e2/me-3gaz_1ub6_1080g1ypnor6yvttsr-big-Data.db:level=2:origin=compaction]
```

**Comment by mykaul** [comment]
- The Nemesis causing it seems the same - ModifyTable in both cases.
- If there's no user visible impact, it's worth noting. Nevertheless, the error is a bit cryptic and is interesting to understand better.

**Comment by mykaul** [comment]
@bhalevy - thoughts?

**Comment by mykaul** [comment]
5.4 rc1 failure - https://argus.scylladb.com/test/f3d43075-df2d-42d4-b609-f0c7a1531f87/runs?additionalRuns[]=e4a990db-f8c7-4437-8fb1-8fb579ff89b1

**Comment by bhalevy** [comment]
This looks like a race between the schema change that adds compression to ongoing sstable writers in compaction that is later resolved by retrying the compactions.

@michoecho can you please look into this issue?

longevity-50gb-12h-5-4-db-node-e4a990db-6/messages.log
```
2023-11-09T06:33:44.599+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:main] init - Scylla version 5.4.0~rc1-0.20231029.24efacf90d0b with build-id f59b16b61ac1631394eebf369635ef29f7918e60 starting ...
...
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:stat] migration_manager - Update table 'keyspace1.standard1' From org.apache.cassandra.config.CFMetaData@0x6000061d5500[cfId=5d79c820-7e36-11ee-9781-f8c917fbbd51,ksName=keyspace1,cfName=standard1,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0,dcLocalReadRepairChance=0,tombstoneGcOptions={"mode":"timeout","propagation_delay_in_seconds":"3600"},gcGraceSeconds=864000,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=C0, type=org.apache.cassandra.db.marshal.BytesType, kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=b706f9bd-894a-3116-b9e3-fd7175c48670,droppedColumns={},collections={},indices={}] To org.apache.cassandra.config.CFMetaData@0x60000575e000[cfId=5d79c820-7e36-11ee-9781-f8c917fbbd51,ksName=keyspace1,cfName=standard1,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0,dcLocalReadRepairChance=0,tombstoneGcOptions={"mode":"timeout","propagation_delay_in_seconds":"3600"},gcGraceSeconds=864000,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=C0, type=org.apache.cassandra.db.marshal.BytesType, kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={chunk_length_in_kb=128, crc_check_chance=0.795583, sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=1780b9b0-7ed5-11ee-a18a-a2b5ef9fbdf9,droppedColumns={},collections={},indices={}]
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:main] schema_tables - Altering keyspace1.standard1 id=5d79c820-7e36-11ee-9781-f8c917fbbd51 version=1ea4aa60-3c99-3579-868f-1da0b36c15b4
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 6:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 8:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 7:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 2:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.735+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 1:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 5:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 3:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 4:main] query_processor - Column definitions for keyspace1.standard1 changed, invalidating related prepared statements
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:main] schema_tables - Schema version changed to c55fd24e-0dd5-3b35-bfb8-84c6c3548076
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6      !ERR | scylla[19052]:  [shard 2:comp] sstable - writer failed to close file: std::runtime_error (compress is not supported)
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 2:comp] compaction - [Compact keyspace1.standard1 13ed59c0-7ed5-11ee-b559-a2b4ef9fbdf9] Compacting of 13 sstables interrupted due to: std::runtime_error (compress is not supported)
2023-11-09T07:53:38.736+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6      !ERR | scylla[19052]:  [shard 2:comp] compaction_manager - Compaction task 0x60400851db00 for table keyspace1.standard1 compaction_group=0/1 [0x6040046dc1e0]: failed: std::runtime_error (compress is not supported). Will retry in 5 seconds
...
2023-11-09T07:53:44.235+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:comp] compaction - [Compact keyspace1.standard1 1ad31900-7ed5-11ee-a18a-a2b5ef9fbdf9] Compacting [/var/lib/scylla/data/keyspace1/standard1-5d79c8207e3611ee9781f8c917fbbd51/me-3gax_0llj_097sg2gfr08d704u7t-big-Data.db:level=2:origin=compaction,/var/lib/scylla/data/keyspace1/standard1-5d79c8207e3611ee9781f8c917fbbd51/me-3gax_0lxd_5l9dc2gfr08d704u7t-big-Data.db:level=2:origin=compaction]
...
2023-11-09T07:53:46.235+00:00 longevity-50gb-12h-5-4-db-node-e4a990db-6     !INFO | scylla[19052]:  [shard 0:comp] compaction - [Compact keyspace1.standard1 1ad31900-7ed5-11ee-a18a-a2b5ef9fbdf9] Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-5d79c8207e3611ee9781f8c917fbbd51/me-3gax_0lxj_4wtqo2gfr08d704u7t-big-Data.db:level=2,/var/lib/scylla/data/keyspace1/standard1-5d79c8207e3611ee9781f8c917fbbd51/me-3gax_0lxl_097sg2gfr08d704u7t-big-Data.db:level=2]. 341MB to 298MB (~87% of original) in 2037ms = 167MB/s. ~316416 total partitions merged to 278360.
```

**Comment by michoecho** [comment]
I spent a long while staring at the code and this doesn't make sense. It seems that it should be impossible for this exception to be thrown:

1. `compress is not supported` is only thrown from `local_compressor::compress`, when `local_compressor` has a null  `_compressor`.
2. `local_compression::_compressor` is never modified, so if it's null, then it must have been constructed that way. It is constructed as a copy of `_schema->get_compressor_params().get_compressor()`.
3. But `local_compression` is *only* constructed when `sstable::_compression_enabled`, and that's only true if `_schema->get_compressor_params().get_compressor() != nullptr`.

So this is a contradiction. I'm stumped.

**Comment by michoecho** [comment]
Hmm, after I typed the above out, I got an idea. There are two different `_schema`s, one in the sstable and one in its writer. There might be no contradiction if we assume that the `_schema` of the sstable is different than the `_schema` of its writer. But is this possible?

**Comment by michoecho** [comment]
> There might be no contradiction if we assume that the _schema of the sstable is different than the _schema of its writer. But is this possible?

Yes, it is.

So the bug is that the sstable writer decides *if* to compress based on the schema of the sstable (which is equal to the table's schema at the moment when the sstable was created), but the actual configuration of the compression is taken from the schema of the writer (equal to the schema of the *reader*/*source*).

So if compression is ALTERed from no compression to some compression in the middle of a compression or flush (before the last output sstable is created) , the writer will attempt to compress (because the schema of the last sstables has compression), but it will fail because its compressor is empty (because that's what the writer's schema says).

What a mess. Why does this have to be so convoluted.

**Comment by mykaul** [comment]
@michoecho - thanks for the explanation. I wonder what other configurations ALTER might do to tables that can be more harmful.

**Comment by bhalevy** [comment]
> > There might be no contradiction if we assume that the _schema of the sstable is different than the _schema of its writer. But is this possible?
> 
> Yes, it is.
> 
> So the bug is that the sstable writer decides _if_ to compress based on the schema of the sstable (which is equal to the table's schema at the moment when the sstable was created), but the actual configuration of the compression is taken from the schema of the writer (equal to the schema of the _reader_/_source_).
> 
> So if compression is ALTERed from no compression to some compression in the middle of a compaction/flush (before the last output sstable is created) , the writer will attempt to compress (because the schema of the last sstables has compression), but it will fail because its compressor is empty (because that's what the writer's schema says).
> 
> What a mess. Why does this have to be so convoluted.

@tgrabiec will probably know the answer to why it was done this way.
My guess is that to enable `ALTER TABLE` while there are writes in progress we take a "snapshot" of the schema when the write starts (using a `schema_ptr`) that is required to format the SSTable when it's written.
The mess seems to be the mixture of the 2.  The writer should stick to the schema it captured.

**Comment by tgrabiec** [comment]
> @tgrabiec will probably know the answer to why it was done this way.

I don't know about the compression logic. One thing is that the writer has to use the same schema as the memtable reader, to properly interpret the fragments it consumes.

**Comment by tgrabiec** [comment]
> What a mess. Why does this have to be so convoluted.

I don't see a reason not to stick to one schema version.

**Comment by denesb** [comment]
We have seen another variant of this: scylla crashing when attempting to write the sstable filter, which was created with fp chance 1.0, but was altered during the write to 0.01.

**Comment by mykaul** [comment]
@denesb - what's the next step here?

**Comment by denesb** [comment]
> @denesb - what's the next step here?

Implementing the fix, which is to stick to the same schema which was used to create the sstable at the start of the write.

**Comment by nopzdk** [comment]
@mykaul @michoecho we looked at the patch with @tarzanek and could not find this fix in any 2024.1.x release notes or git
